### PR TITLE
Convert MQTT server user preference to hostname:port

### PR DESCRIPTION
**Pull request by thebentern**
I'm not purveyor of idiomatic C++ / Arduino code by any means, but based on your suggestion, I threw together a change for splitting MQTT host and port.
Feel free to kill it or change it or toss it back 😎 
I suppose we'll also need to ensure that this is documented in the user preference docs.

### CODE REVIEW

**Comment by geeksville** [comment]
This pull request has been mentioned on **Meshtastic**. There might be relevant details there:

https://meshtastic.discourse.group/t/new-alpha-1-2-25-to-test-the-mqtt-gateway-feature/3106/50

File: src/mqtt/MQTT.cpp [hunk 1]
```diff
@@ -62,12 +62,17 @@ MQTT::MQTT() : concurrency::OSThread("mqtt"), pubSub(mqttClient)
...
     // pubSub.setServer("devsrv.ezdevice.net", 1883); or 192.168.10.188
-    const char *serverAddr = "mqtt.meshtastic.org"; // default hostname
+    const char *serverAddr = "mqtt.meshtastic.org:1883"; // default hostname
 
     if (*radioConfig.preferences.mqtt_server)
         serverAddr = radioConfig.preferences.mqtt_server; // Override the default
 
-    pubSub.setServer(serverAddr, 1883);
+    String server = String(serverAddr);  
+    int delimIndex = server.indexOf(':');
```
Comment (jdstroy) [inline src/mqtt/MQTT.cpp:71]: Does this actually build?  I am not aware of indexOf or substring being in the C++ standard library.

File: src/mqtt/MQTT.cpp [hunk 2]
```diff
@@ -62,12 +62,17 @@ MQTT::MQTT() : concurrency::OSThread("mqtt"), pubSub(mqttClient)
...
+    const char *serverAddr = "mqtt.meshtastic.org:1883"; // default hostname
 
     if (*radioConfig.preferences.mqtt_server)
         serverAddr = radioConfig.preferences.mqtt_server; // Override the default
 
-    pubSub.setServer(serverAddr, 1883);
+    String server = String(serverAddr);  
+    int delimIndex = server.indexOf(':'); 
+    String host = server.substring(0, delimIndex);
+    String port = server.substring(delimIndex+1, server.length());
```
Comment (jdstroy) [inline src/mqtt/MQTT.cpp:73]: This algorithm will fail when `serverAddr` does not contain ':'.  The default port should be 1883 if delilmIndex isn't valid.

**Review by jdstroy — CHANGES_REQUESTED** [review]

File: src/mqtt/MQTT.cpp (hunk 2 above)
Comment (thebentern) [reply src/mqtt/MQTT.cpp]: Good call. My initial thought was that we would just document the format, but this will be more user friendly.

**Review by jdstroy — APPROVED** [review]

**Comment by geeksville** [comment]
super great - sorry for the delay, I'll merge tomorrow.

**Comment by thebentern** [comment]
Thanks, Kevin. Good to see you back!

**Comment by flux242** [comment]
you are referencing a deallocated string. Connection works only by a coincidence

**Comment by flux242** [comment]
please consider the following patch
```diff
diff --git a/src/mqtt/MQTT.cpp b/src/mqtt/MQTT.cpp
index 33e0ee7..450c844 100644
--- a/src/mqtt/MQTT.cpp
+++ b/src/mqtt/MQTT.cpp
@@ -62,7 +62,8 @@ MQTT::MQTT() : concurrency::OSThread("mqtt"), pubSub(mqttClient)
 void MQTT::reconnect()
 {
     // pubSub.setServer("devsrv.ezdevice.net", 1883); or 192.168.10.188
-    const char *serverAddr = "mqtt.meshtastic.org:1883"; // default hostname
+    const char *serverAddr = "mqtt.meshtastic.org"; // default hostname
+    int serverPort = 1883; // default server port
 
     if (*radioConfig.preferences.mqtt_server)
         serverAddr = radioConfig.preferences.mqtt_server; // Override the default
@@ -70,15 +71,16 @@ void MQTT::reconnect()
     String server = String(serverAddr);  
     int delimIndex = server.indexOf(':'); 
     if (delimIndex > 0) {
-        String host = server.substring(0, delimIndex);
         String port = server.substring(delimIndex+1, server.length());
-        pubSub.setServer(host.c_str(), port.toInt());
+        server[delimIndex] = 0;
+        serverPort = port.toInt();
+        pubSub.setServer(server.c_str(), serverPort);
     }
     else {
-        pubSub.setServer(serverAddr, 1883);
+        pubSub.setServer(serverAddr, serverPort);
     }
 
-    DEBUG_MSG("Connecting to MQTT server\n", serverAddr);
+    DEBUG_MSG("Connecting to MQTT server %s, port: %d\n", server.c_str(), serverPort);
     auto myStatus = (statusTopic + owner.id);
     bool connected = pubSub.connect(owner.id, "meshdev", "large4cats", myStatus.c_str(), 1, true, "offline");
     if (connected) {
```

**Comment by sachaw** [comment]
@flux242 Could you please open a new PR

**Comment by geeksville** [comment]
oh yes @flux242 that is a GREAT fix and I'm super loaded right now so if you could send in a PR it would be a big help.

**Comment by geeksville** [comment]
I'm working on code today so I'll make a fix based on @flux242 great comment

**Comment by flux242** [comment]
sorry, my patch wasn't quite correct. Here is an updated version
```diff
diff --git a/src/mqtt/MQTT.cpp b/src/mqtt/MQTT.cpp
index 33e0ee7..d4a9b85 100644
--- a/src/mqtt/MQTT.cpp
+++ b/src/mqtt/MQTT.cpp
@@ -62,7 +62,8 @@ MQTT::MQTT() : concurrency::OSThread("mqtt"), pubSub(mqttClient)
 void MQTT::reconnect()
 {
     // pubSub.setServer("devsrv.ezdevice.net", 1883); or 192.168.10.188
-    const char *serverAddr = "mqtt.meshtastic.org:1883"; // default hostname
+    const char *serverAddr = "mqtt.meshtastic.org"; // default hostname
+    int serverPort = 1883; // default server port
 
     if (*radioConfig.preferences.mqtt_server)
         serverAddr = radioConfig.preferences.mqtt_server; // Override the default
@@ -70,15 +71,13 @@ void MQTT::reconnect()
     String server = String(serverAddr);  
     int delimIndex = server.indexOf(':'); 
     if (delimIndex > 0) {
-        String host = server.substring(0, delimIndex);
         String port = server.substring(delimIndex+1, server.length());
-        pubSub.setServer(host.c_str(), port.toInt());
-    }
-    else {
-        pubSub.setServer(serverAddr, 1883);
+        server[delimIndex] = 0;
+        serverPort = port.toInt();
     }
+    pubSub.setServer(server.c_str(), serverPort);
 
-    DEBUG_MSG("Connecting to MQTT server\n", serverAddr);
+    DEBUG_MSG("Connecting to MQTT server %s, port: %d\n", server.c_str(), serverPort);
     auto myStatus = (statusTopic + owner.id);
     bool connected = pubSub.connect(owner.id, "meshdev", "large4cats", myStatus.c_str(), 1, true, "offline");
     if (connected) {
```

**Pull request merged**